File: go/test/timezone_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/world-time-sdk/go"
	"github.com/voxgig-sdk/world-time-sdk/go/core"

	vs "github.com/voxgig-sdk/world-time-sdk/go/utility/struct"
)

func TestTimezoneEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Timezone(nil)
		if ent == nil {
			t.Fatal("expected non-nil TimezoneEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := timezoneBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "timezone." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set WORLDTIME_TEST_TIMEZONE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		timezoneRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.timezone", setup.data)))
		var timezoneRef01Data map[string]any
		if len(timezoneRef01DataRaw) > 0 {
			timezoneRef01Data = core.ToMapAny(timezoneRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = timezoneRef01Data

		// LIST
		timezoneRef01Ent := client.Timezone(nil)
		timezoneRef01Match := map[string]any{}

		timezoneRef01ListResult, err := timezoneRef01Ent.List(timezoneRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, timezoneRef01ListOk := timezoneRef01ListResult.([]any)
		if !timezoneRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", timezoneRef01ListResult)
		}

		// LOAD
		timezoneRef01MatchDt0 := map[string]any{}
		timezoneRef01DataDt0Loaded, err := timezoneRef01Ent.Load(timezoneRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if timezoneRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func timezoneBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "timezone", "TimezoneTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read timezone test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse timezone test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"timezone01", "timezone02", "timezone03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("WORLDTIME_TEST_TIMEZONE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"WORLDTIME_TEST_TIMEZONE_ENTID": idmap,
		"WORLDTIME_TEST_LIVE":      "FALSE",
		"WORLDTIME_TEST_EXPLAIN":   "FALSE",
		"WORLDTIME_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["WORLDTIME_TEST_TIMEZONE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["WORLDTIME_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["WORLDTIME_APIKEY"],
			},
			extra,
		})
		client = sdk.NewWorldTimeSDK(core.ToMapAny(mergedOpts))
	}

	live := env["WORLDTIME_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["WORLDTIME_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/timezone_entity_test.rb
# Timezone entity test

require "minitest/autorun"
require "json"
require_relative "../WorldTime_sdk"
require_relative "runner"

class TimezoneEntityTest < Minitest::Test
  def test_create_instance
    testsdk = WorldTimeSDK.test(nil, nil)
    ent = testsdk.Timezone(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "timezone" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = WorldTimeSDK.test(seed, nil)
    seen = base.Timezone(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = WorldTimeConfig.shared_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = WorldTimeSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.Timezone(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = timezone_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "timezone." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set WORLD_TIME_TEST_TIMEZONE_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    timezone_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.timezone")))
    timezone_ref01_data = nil
    if timezone_ref01_data_raw.length > 0
      timezone_ref01_data = Helpers.to_map(timezone_ref01_data_raw[0][1])
    end

    # LIST
    timezone_ref01_ent = client.Timezone(nil)
    timezone_ref01_match = {}

    timezone_ref01_list_result = timezone_ref01_ent.list(timezone_ref01_match, nil)
    assert timezone_ref01_list_result.is_a?(Array)

    # LOAD
    timezone_ref01_match_dt0 = {}
    timezone_ref01_data_dt0_loaded = timezone_ref01_ent.load(timezone_ref01_match_dt0, nil)
    assert !timezone_ref01_data_dt0_loaded.nil?

  end
end

def timezone_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "timezone", "TimezoneTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = WorldTimeSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["timezone01", "timezone02", "timezone03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["WORLD_TIME_TEST_TIMEZONE_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "WORLD_TIME_TEST_TIMEZONE_ENTID" => idmap,
    "WORLD_TIME_TEST_LIVE" => "FALSE",
    "WORLD_TIME_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["WORLD_TIME_TEST_TIMEZONE_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["WORLD_TIME_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = WorldTimeSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["WORLD_TIME_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["WORLD_TIME_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
